martian/src/lib.rs

```diff
@@ -126,46 +81,114 @@ fn setup_logging(log_file: &File, level: LevelFilter) {
     }
 }
 
-pub fn martian_main<S: ::std::hash::BuildHasher>(
-    args: Vec<String>,
+pub struct MartianAdapter<S> {
     stage_map: HashMap<String, Box<dyn RawMartianStage>, S>,
-) -> Result<(), Error> {
-    martian_main_with_log_level(args, stage_map, LevelFilter::Debug)
+    log_level: LevelFilter,
+    is_error_assert: Box<dyn (Fn(&Error) -> bool) + 'static>,
 }
 
-pub fn martian_main_with_log_level<S: ::std::hash::BuildHasher>(
+impl<S: std::hash::BuildHasher> MartianAdapter<S> {
+    pub fn new(stage_map: HashMap<String, Box<dyn RawMartianStage>, S>) -> MartianAdapter<S> {
+        MartianAdapter {
+            stage_map,
+            log_level: LevelFilter::Warn,
+            is_error_assert: Box::new(|_| false),
+        }
+    }
+
+    pub fn log_level(self, log_level: LevelFilter) -> MartianAdapter<S> {
+        MartianAdapter {
+            log_level,
+            .. self
+        }
+    }
+
+    pub fn assert_if<F: 'static + Fn(&Error) -> bool>(self, predicate: F) -> MartianAdapter<S> {
+        MartianAdapter {
+            is_error_assert: Box::new(predicate),
+            .. self
+        }
+    }
+
+    pub fn run(self, args: Vec<String>) -> (i32, Option<Error>) {
+        martian_entry_point(
+            args,
+            self.stage_map,
+            self.log_level,
+            self.is_error_assert)
+    }
+}
+
+
+/// Run the martian adapter using the given cmdline args
+/// provided by the martian runtime. If there is an error
+/// in the stage setup, this returns Err(e). If the stage
+/// is executed it returns `Ok((returncode, option_error))`. 
+/// The caller should call sys::exit() with the returncode. 
+/// If the stage itself failed, the error causing the failure 
+/// will be returned in the `option_error`.
+/// Arguments:
+///  - `args`: vector of command line arguments, typicall supplied by Martian runtime.
+///  - `stage_map`: names and implementations of the Martian stages that can be run by this binary.
+///  - `level`: the level of log messages that are emitted to the _log file
+///  - `is_error_assert`: a predicate determining whether to emit an error as an ASSERT
+///    to Martian. ASSERT errors indicate an unrecoverable configuration error, and will
+///    prevent the user from restarting the pipeline. The is_error_assert function should 
+///    use downcasting to match the error against a set of error types that should generate an assert.
+fn martian_entry_point<S: std::hash::BuildHasher>(
     args: Vec<String>,
     stage_map: HashMap<String, Box<dyn RawMartianStage>, S>,
     level: LevelFilter,
-) -> Result<(), Error> {
+    is_error_assert: Box<dyn Fn(&Error) -> bool>,
+) -> (i32, Option<Error>) {
```

Comment:
Why is this returning `Option<Error>`?  It makes it look like the caller is ignoring an error, but the error has already been reported at that point.  Doesn't look like anything ever uses that return value either.

Also doesn't rust have a way of specifying that users shouldn't ignore a given return value?  It's probably a bad idea to ignore the return code.

Reply:
good call - adding a `#[must_use]` and making the primary version of this API only return the retcode. I'll add another method that will give you the error (if any) that I want to have for unit testing.